feat: Ability to resend registration email

huglx opened the pull request:
Added: 
- limit the number of the requests per hour to some constant per e-mail and ip
- Enable to the user to login, but without any permissions, telling them, they should verify
- backend endpoint to resend the verification link
Hi @JanCizmar, could you please provide feedback so I can see if I am doing this correctly? Thanks

JanCizmar commented on backend/security/src/main/kotlin/io/tolgee/security/authorization/OrganizationAuthorizationInterceptor.kt:
```diff
@@ -52,7 +56,16 @@ class OrganizationAuthorizationInterceptor(
     handler: HandlerMethod,
   ): Boolean {
-    val userId = authenticationFacade.authenticatedUser.id
+    val user = authenticationFacade.authenticatedUserEntity
```
Don't use entity here. We don't want to fetch user entity from the DB with every request. Thats why authenticatedUser is DTO, which can be returned from cache.

JanCizmar commented on backend/security/src/main/kotlin/io/tolgee/security/authorization/OrganizationAuthorizationInterceptor.kt:
```diff
@@ -52,7 +56,16 @@ class OrganizationAuthorizationInterceptor(
-    val userId = authenticationFacade.authenticatedUser.id
+    val user = authenticationFacade.authenticatedUserEntity
+
+    if (!emailVerificationService.isVerified(
```
This code is duplicated, across the codebase. Create some method for this.

JanCizmar commented on backend/data/src/main/kotlin/io/tolgee/service/EmailVerificationService.kt:
```diff
@@ -63,6 +66,49 @@ class EmailVerificationService(
+  }
+
+  @Transactional
+  fun fakeSend(
```
This method us unused.

JanCizmar commented on backend/data/src/main/kotlin/io/tolgee/service/EmailVerificationService.kt:
```diff
@@ -63,6 +66,49 @@ class EmailVerificationService(
   }
 
+  @Transactional
+  fun resendEmailVerification(
```
I think, maybe strict `@RateLimited` annotation on the endpoints would be enough and simplier.

JanCizmar replied:
Tell me if I am wrong.

huglx replied:
I think you're right, but I wanted to solve the problem with incorrect email(more in conversation below).

JanCizmar commented on backend/data/src/main/kotlin/io/tolgee/security/ratelimit/RateLimitService.kt:
```diff
@@ -139,6 +139,23 @@ class RateLimitService(
+    if (!rateLimitProperties.emailVerificationRequestLimitEnabled || email.isNullOrEmpty()) return null
+
+    val ip = request.remoteAddr
+    val key = "global.ip.$ip.email.$email::auth"
```
So If I resend it from other IP, I can bypass the rate limit? Why so?

huglx replied:
I did this because I thought that if a user entered the wrong email and sent the verification N times (so he has a rate limit), he could change the email and send the verification more times, but you correctly noticed the problem with different IPs

huglx commented:
I added the FE part. @stepan662, please review it. And I have some questions on the FE part. I added icons in 2 variants (dark mode and light mode), is this a normal way to store multiple icons? And I wrote a part in ApiErrorHandling that handles 429 error, check if it's ok.

I also have a question about the speed limiter. Is it OK that I only store the IP of the user? That way, if the user entered the wrong email and sent too many requests, he has to wait for a while after entering the correct email. On the other hand, someone could change the IP and spam.

stepan662 commented on webapp/src/svgs/icons/mailDark.svg:
Use fill="currentColor" in the icon and then you can set the color from the code rather having two separate icons

stepan662 commented on webapp/src/views/projects/ProjectListView.tsx:
I think this is overly complex, could we have two separate files like `ProjectListView` and `EmailNotVerifiedView`. There is a router rule, that it redirects you to the page `/projects` after login, so you can change that in case the email is not verified. Or if thats too complicated, you can just switch that directly in the router file.

stepan662 commented on webapp/src/component/layout/TopBanner/TopBanner.tsx:
```diff
@@ -17,9 +17,15 @@ const StyledContainer = styled('div')`
   width: 100%;
-  background: ${({ theme }) => theme.palette.topBanner.background};
+  background: ${({ theme }) =>
+    useIsEmailVerified()
```
This is not a good way how to use hooks. Use it in the component itself and only apply class to change styles.

https://react.dev/reference/rules/rules-of-hooks

stepan662 commented on webapp/src/i18n/en.json:
```diff
@@ -1399,5 +1399,14 @@
+  "verify_email_description" : "The final step is to verify your email. After verification, you will be able to use the platform.\nPlease check your inbox at <b>{email}</b>.",
+  "verify_email_now_title" : "Verify your email now",
+  "verify_email_resend_button" : "resend verification email",
+  "verify_email_title" : "Verify your email address"
```
You don't have to edit i18n files manually, use in-context tools and create the translations in the tolgee platform, i18n files are then updated automatically in the release pipeline

stepan662 commented on webapp/src/service/http/ApiHttpService.tsx:
```diff
@@ -59,8 +59,9 @@ export class ApiHttpService {
+              const currentPath = location.pathname;
               resultError.setErrorHandler(() =>
-                handleApiError(r, responseData, init, options)
+                handleApiError(r, responseData, init, options, currentPath)
```
It would be nice to get rid of this (as mentioned before)

stepan662 commented on webapp/src/colors.tsx:
```diff
@@ -306,6 +317,11 @@ export const colors = {
+      background: '#822343',
+      mainText: '#FDECF2',
+      titleText: '#F6A7C2',
+    },
```
This way of defining colors is deprecated, we now use tokens directly exported from figma (in `figmaTheme.ts`). If you check what colors are used in figma, the same colors should be available in `theme.palette.tokens...`. If there are new colors, which are not in the file, you need to use our figma plugin to update the palette https://github.com/tolgee/figma-color-extractor (I can help you with that if needed).

huglx replied:
aha, ok

stepan662 commented on e2e/cypress/e2e/security/signUp.cy.ts:
```diff
@@ -133,6 +134,16 @@ context('Sign up', () => {
+    gcy('resend-email-button').click();
+    cy.contains('Your verification link has been resent.');
+  });
+
```
We should also check that verification link works (if it's not tested somewhere else)

stepan662 commented on webapp/src/component/CustomIcons.tsx:
```diff
@@ -40,6 +42,12 @@ export const ImportIcon: React.FC<IconProps> = (props) => (
+export const MailIcon: React.FC<IconProps> = (props) => (
+  <CustomIcon icon={MailSvg} {...props} />
+);
+export const MailIconDark: React.FC<IconProps> = (props) => (
```
This is not icon, it shoudn't be here.

huglx replied:
Why not? Then how can I use them?

huglx replied:
Should I use this like image?

stepan662 commented on webapp/src/component/EmailNotVerifiedView.tsx:
```diff
@@ -0,0 +1,117 @@
+            </StyledDescription>
+            <StyledImg src={imageSrc} alt="Verify email" />
+            <StyledHint variant="body2">
+              <T keyName="didnt_receive_email_hint" />
```
It's good to have some common prefix, like here I'd prefix everything with `verify_email_`, so it's better organized

huglx commented on webapp/src/service/http/handleApiError.tsx:
```diff
@@ -29,6 +30,16 @@ export const handleApiError = (
+      if (resObject?.code === 'email_not_verified') {
+        if (emailVerificationTemplate.test(currentPath)) return;
+
+        globalContext.actions?.redirectToUnverifiedView();
```
There's an issue that the global context always sends requests (even if it's login/verify/ link). So without this check, when you click on the verify link, it redirects to an unverified view.

huglx replied:
You're right. But when the user clicks on the verification link, it redirect him to the EmailVerificationHandler, but the global context calls the protected endpoind and throws a 403. this check looks if path is /login/verify_email/{id}/{code}, so the error handler doesn't redirect the user to the unverifiedView.

stepan662 replied:
So, can we prevent the global context from calling the unprotected endpoint?

stepan662 commented on webapp/src/component/RootRouter.tsx:
```diff
@@ -122,6 +122,9 @@ export const RootRouter = () => (
       <Switch>
+        <PrivateRoute exact path={LINKS.PROJECTS.template}>
+          <Redirect to={LINKS.ROOT.template} />
+        </PrivateRoute>
```
Cool 👍

stepan662 commented on webapp/src/component/RootRouter.tsx:
```diff
@@ -122,6 +122,9 @@ export const RootRouter = () => (
+        </PrivateRoute>
         <PrivateRoute path={LINKS.PROJECTS.template}>
           <ProjectsRouter />
         </PrivateRoute>
```
This can be removed, no?

stepan662 commented:
Ok, seems like everything is working, but there are some e2e tests failing.
